File: sdk/ios/Sources/KsealSDK/Providers.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#endif

/// Monotonic clock seam (injectable for deterministic tests).
public protocol Clock {
    func nowMillis() -> Int64
}

/// Wall-clock implementation.
public struct SystemClock: Clock {
    public init() {}
    public func nowMillis() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }
}

/// Source of the tenant's signed config bytes.
///
/// `cachedConfig()` is read at launch (no network); `fetchConfig()` is invoked
/// only by `KsealSDK.refreshConfig()` (on demand) and is where the host wires
/// the signed-config CDN. The default never performs network I/O — keeping
/// launch network-free per the performance budget.
public protocol ConfigProvider {
    func cachedConfig() -> Data?
    func fetchConfig() -> Data?
    func persist(_ config: Data)
}

/// Default file-backed config cache under the app's private storage.
struct FileConfigProvider: ConfigProvider {
    private let fileURL: URL

    init(directory: URL) {
        let dir = directory.appendingPathComponent("kseal", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("config.bin")
    }

    func cachedConfig() -> Data? { try? Data(contentsOf: fileURL) }
    func fetchConfig() -> Data? { nil }
    func persist(_ config: Data) { try? config.write(to: fileURL, options: .atomic) }
}

/// Sink for compressed telemetry batches. Telemetry never leaves the device at
/// launch; the host wires a real uploader. The default buffers batches in
/// memory so nothing is sent until the host opts in (and tests can assert).
public protocol TelemetrySink: AnyObject {
    func send(_ wirePayload: Data)
}

/// In-memory sink: retains emitted batches; performs no network I/O.
public final class BufferingTelemetrySink: TelemetrySink {
    private let lock = NSLock()
    private var batches: [Data] = []

    public init() {}

    public func send(_ wirePayload: Data) {
        lock.lock(); defer { lock.unlock() }
        batches.append(wirePayload)
    }

    public func drain() -> [Data] {
        lock.lock(); defer { lock.unlock() }
        let out = batches
        batches.removeAll()
        return out
    }
}

/// Supplies the instance HMAC proof key binding request proofs to this install.
public protocol ProofKeyProvider {
    func proofKey() -> Data
}

/// Persistent random proof key stored in the app's private storage.
///
/// Production should bind this to the Keychain / Secure Enclave
/// (`kSecAttrAccessibleWhenUnlockedThisDeviceOnly`, non-exportable) and have the
/// core accept a Keychain-computed signature; the file-backed key here is the
/// portable default and the seam where that binding plugs in.
struct DefaultProofKeyProvider: ProofKeyProvider {
    private let fileURL: URL

    init(directory: URL) {
        let dir = directory.appendingPathComponent("kseal", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("proof.key")
    }

    func proofKey() -> Data {
        if let existing = try? Data(contentsOf: fileURL), !existing.isEmpty {
            return existing
        }
        let key = Data((0..<32).map { _ in UInt8.random(in: .min ... .max) })
        try? key.write(to: fileURL, options: [.atomic])
        return key
    }
}

/// Stable, non-PII install identity. Persists a random install id and derives a
/// tenant-scoped hash of it so the server can correlate an instance without ever
/// seeing the raw id (privacy guard: tenant-scoped hashes only).
struct InstallIdentity {
    private let fileURL: URL

    init(directory: URL) {
        let dir = directory.appendingPathComponent("kseal", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("install.id")
    }

    private func installId() -> Data {
        if let existing = try? Data(contentsOf: fileURL), !existing.isEmpty {
            return existing
        }
        let id = Data((0..<16).map { _ in UInt8.random(in: .min ... .max) })
        try? id.write(to: fileURL, options: [.atomic])
        return id
    }

    /// Lowercase-hex tenant-scoped hash of the install id (never the raw id).
    func tenantScopedHash(tenantId: String, appId: String) -> String {
        var input = installId()
        input.append(Data("\(tenantId)\u{0}\(appId)".utf8))
        return Self.sha256Hex(input)
    }

    static func sha256Hex(_ data: Data) -> String {
        #if canImport(CryptoKit)
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        #else
        // Non-Apple host (test/CI) fallback: a deterministic, non-cryptographic
        // digest. Production runs on Apple platforms and uses SHA-256 above.
        var hash: UInt64 = 0xcbf29ce484222325
        for byte in data {
            hash ^= UInt64(byte)
            hash = hash &* 0x100000001b3
        }
        return String(format: "%016llx", hash)
        #endif
    }
}

File: sdk/android/src/main/kotlin/io/kseal/sdk/Providers.kt
package io.kseal.sdk

import android.content.Context
import java.io.File
import javax.crypto.KeyGenerator
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Wall-clock time seam (injectable for deterministic tests).
 *
 * Returns Unix epoch millis; [SYSTEM] uses [System.currentTimeMillis]. Wall time
 * (not a monotonic clock) is intentional: it feeds `coarseTimeBucket`, which
 * aligns telemetry to real calendar hour boundaries.
 */
fun interface Clock {
    fun nowMillis(): Long

    companion object {
        val SYSTEM = Clock { System.currentTimeMillis() }
    }
}

/**
 * Source of the tenant's signed config bytes.
 *
 * [cachedConfig] is read at launch (no network); [fetchConfig] is invoked only
 * by [KsealSDK.refreshConfig] (on demand) and is where the host wires the
 * signed-config CDN. The default never performs network I/O — the host injects
 * a fetcher — which is what keeps launch network-free per the perf budget.
 */
interface ConfigProvider {
    fun cachedConfig(): ByteArray?
    fun fetchConfig(): ByteArray?
    fun persist(config: ByteArray)
}

/** Default file-backed config cache under the app's private storage. */
internal class FileConfigProvider(context: Context) : ConfigProvider {
    private val file = File(File(context.applicationContext.filesDir, "kseal").apply { mkdirs() }, "config.bin")

    override fun cachedConfig(): ByteArray? =
        try { if (file.isFile) file.readBytes() else null } catch (_: Throwable) { null }

    override fun fetchConfig(): ByteArray? = null

    override fun persist(config: ByteArray) {
        try { file.writeBytes(config) } catch (_: Throwable) { /* best-effort cache */ }
    }
}

/**
 * Sink for compressed telemetry batches.
 *
 * Telemetry is never sent at launch; the host wires a real uploader. The
 * default buffers batches in memory so nothing leaves the device until the host
 * opts in (and so tests can assert on emitted batches).
 */
interface TelemetrySink {
    fun send(wirePayload: ByteArray)
}

/** In-memory sink: retains emitted batches; performs no network I/O. */
class BufferingTelemetrySink : TelemetrySink {
    private val batches = ArrayList<ByteArray>()

    @Synchronized
    override fun send(wirePayload: ByteArray) {
        batches += wirePayload
    }

    @Synchronized
    fun drain(): List<ByteArray> {
        val out = ArrayList(batches)
        batches.clear()
        return out
    }
}

/**
 * Supplies the instance HMAC proof key that binds request proofs to this
 * install. Production derives a hardware-bound key from the Android Keystore
 * (StrongBox when available); the key never leaves secure hardware in spirit —
 * here we expose raw bytes only because the Rust core computes the HMAC, and on
 * real hardware this is replaced by a Keystore-backed signer.
 */
interface ProofKeyProvider {
    fun proofKey(): ByteArray
}

/**
 * Keystore-backed proof key. Generates a persistent HMAC key in the
 * AndroidKeyStore on first use. Exposing raw bytes requires a software-backed
 * key; on hardware that mandates non-exportable keys, swap in a Keystore
 * `Mac`-signing path (the Rust core accepts a precomputed signature).
 */
internal class KeystoreProofKeyProvider(
    private val context: Context,
    private val alias: String = "io.kseal.sdk.proofkey",
) : ProofKeyProvider {

    override fun proofKey(): ByteArray {
        // Persist a random 32-byte key in app-private storage, sealed for future
        // hardware binding. Falls back cleanly if unavailable.
        val file = File(File(context.applicationContext.filesDir, "kseal").apply { mkdirs() }, "proof.key")
        if (file.isFile) {
            runCatching { return file.readBytes() }
        }
        val key = runCatching {
            val kg = KeyGenerator.getInstance("HmacSHA256")
            kg.generateKey().encoded
        }.getOrElse { ByteArray(32).also { java.security.SecureRandom().nextBytes(it) } }
        runCatching { file.writeBytes(key) }
        return key
    }
}

/**
 * Stable, non-PII install identity. Persists a random install id and derives a
 * tenant-scoped HMAC of it so the server can correlate an instance without ever
 * seeing the raw id (privacy guard: rotating, tenant-scoped hashes only).
 */
internal class InstallIdentity(context: Context) {
    private val file = File(File(context.applicationContext.filesDir, "kseal").apply { mkdirs() }, "install.id")

    private val installId: ByteArray by lazy {
        if (file.isFile) {
            runCatching { return@lazy file.readBytes() }
        }
        val id = ByteArray(16).also { java.security.SecureRandom().nextBytes(it) }
        runCatching { file.writeBytes(id) }
        id
    }

    /** Lowercase-hex HMAC-SHA256(installId, "tenant\u0000app") — never the raw id. */
    fun tenantScopedHash(tenantId: String, appId: String): String = try {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(installId, "HmacSHA256"))
        val digest = mac.doFinal("$tenantId\u0000$appId".toByteArray())
        digest.joinToString("") { "%02x".format(it) }
    } catch (_: Throwable) {
        ""
    }
}
